apollo-router-core/src/query_planner/mod.rs

```diff
@@ -395,48 +356,108 @@ impl FetchNode {
+        request: &'a Arc<Request>,
+        service_registry: Arc<dyn ServiceRegistry>,
+        schema: &'a Arc<Schema>,
+    ) -> Result<Value, FetchError> {
+        let FetchNode {
+            operation,
+            service_name,
+            ..
+        } = self;
 
-                    let span = tracing::trace_span!("response_insert");
-                    let _guard = span.enter();
-                    response.append_errors(&mut errors);
-                    response.insert_data(current_dir, data)?;
+        let query_span = tracing::info_span!("subfetch", service = service_name.as_str());
 
-                    Ok(())
-                }
-                None => Err(FetchError::SubrequestNoResponse {
+        let Variables { variables, paths } =
+            self.make_variables(data, current_dir, request, schema)?;
+
+        // We already checked that the service exists during planning
+        let fetcher = service_registry.get(service_name).unwrap();
+
+        let (res, _tail) = fetcher
+            .stream(
+                Request::builder()
+                    .query(operation)
+                    .variables(Some(Arc::new(variables)))
+                    .build(),
+            )
+            .await
+            .into_future()
+            .instrument(query_span)
+            .await;
+
+        let subgraph_response = match res {
+            Some(response) if !response.is_primary() => {
+                return Err(FetchError::SubrequestUnexpectedPatchResponse {
+                    service: service_name.to_owned(),
+                });
+            }
+            Some(subgraph_response) => subgraph_response,
+            None => {
+                return Err(FetchError::SubrequestNoResponse {
                     service: service_name.to_string(),
-                }),
+                })
             }
+        };
+
+        self.response_at_path(current_dir, paths, subgraph_response)
+    }
+
+    fn response_at_path<'a>(
+        &'a self,
+        current_dir: &'a Path,
+        paths: Option<Vec<Path>>,
+        subgraph_response: Response,
+    ) -> Result<Value, FetchError> {
+        let Response { data, .. } = subgraph_response;
+
+        if self.requires.is_some() {
+            // we have to nest conditions and do early returns here
+            // because we need to take ownership of the inner value
+            if let Value::Object(mut map) = data {
+                if let Some(entities) = map.remove("_entities") {
+                    tracing::trace!(
+                        "Received entities: {}",
+                        serde_json::to_string(&entities).unwrap(),
+                    );
+
+                    if let Value::Array(array) = entities {
+                        let span = tracing::trace_span!("response_insert");
+                        let _guard = span.enter();
+
+                        let mut value = Value::default();
+
+                        let paths = paths.unwrap();
```

Comment:
:screwdriver: Please don't use unwrap unless it is very obvious why you are unwrapping. Use `.expect("...")` and write a solid proof that this expect won't *ever* panic.

On top of that, it seems you don't need your field `paths` in the struct `Variables` to be an option at all. All collections in Rust have a function to check for emptiness (often called `is_empty()`). Unless you need to distinguish between `None` and `Some(empty_collection)`, you shouldn't wrap your collection into an Option. That is what I'm fixing in #257

Reply:
makes sense, I'll change that

Reply:
done in 19cae6c